### PR TITLE
Enable gcov

### PR DESCRIPTION
**Pull request by KFilipek**
### Description
This pull request introduces support for gcov, a test coverage program that developers can use to assess the effectiveness of their test suites. The changes include:

- An additional CMake option USE_GCOV to enable gcov support, which is turned off by default.
- Modifications to CMakeLists.txt and cmake/helpers.cmake to ensure that the correct compile and link options are set when USE_GCOV is enabled. It also includes a check to ensure that gcov is only used with Debug builds.
- Updates to CONTRIBUTING.md to guide contributors on how to use gcov to check code coverage for new functionality, including a sample workflow for generating a coverage report.
- Documentation in README.md has been updated to include the new USE_GCOV option in the list of CMake options.

With these changes, contributors can now enable gcov support in their debug builds to generate coverage reports, helping to maintain and improve the quality of the test suite.

### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [ ] CI workflows execute properly

- [x] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [ ] New tests added, especially if they will fail without my changes
- [x] Added/extended example(s) to cover this functionality
- [x] Extended the README/documentation
- [ ] All newly added source files have a license
- [ ] All newly added source files are referenced in CMake files
- [ ] Logger (with debug/info/... messages) is used

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -202,3 +202,18 @@ New dependency: dependency_name
 license: SPDX license tag
 origin: https://dependency_origin.com
 ```
+## Code coverage
```
Comment (bratpiorka) [inline CONTRIBUTING.md:205]: please add an empty line before this section

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -202,3 +202,18 @@ New dependency: dependency_name
...
+
+After adding new functionality add tests and check coverage before and after change.
+To do that, enable coverage instrumentation by enabling USE_GCOV flag in CMake.
+Example flow may look like:
```
Comment (bratpiorka) [inline CONTRIBUTING.md:209]: "To do this, enable coverage instrumentation by turning on the USE_GCOV flag in CMake.
An example flow might look like the following:"

File: README.md [hunk 3]
```diff
@@ -123,6 +123,7 @@ List of options provided by CMake:
 | USE_TSAN | Enable ThreadSanitizer checks | ON/OFF | OFF |
 | USE_MSAN | Enable MemorySanitizer checks | ON/OFF | OFF |
 | USE_VALGRIND | Enable Valgrind instrumentation | ON/OFF | OFF |
+| USE_GCOV | Enable gcov support | ON/OFF | OFF |
```
Comment (bratpiorka) [inline README.md:126]: please add a note somewhere that this will work only for linux

File: CONTRIBUTING.md (hunk 1 above)
Comment (KFilipek) [reply CONTRIBUTING.md:205]: Done

File: CONTRIBUTING.md (hunk 2 above)
Comment (KFilipek) [reply CONTRIBUTING.md:209]: Done

File: README.md (hunk 3 above)
Comment (KFilipek) [reply README.md:126]: Done

File: cmake/helpers.cmake [hunk 4]
```diff
@@ -104,6 +104,12 @@ function(add_umf_target_compile_options name)
                 ${name} PRIVATE -Werror -fno-omit-frame-pointer
                                 -fstack-protector-strong)
         endif()
+        if(USE_GCOV)
+            if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+                message(FATAL_ERROR "To use gcov, the build type must be Debug")
```
Comment (lukaszstolarczuk) [inline cmake/helpers.cmake:109]: is it truly a fatal error? I guess we get a different results for Release, but it should work, right...?

File: cmake/helpers.cmake [hunk 5]
```diff
@@ -121,6 +127,13 @@ function(add_umf_target_link_options name)
     if(NOT MSVC)
         if(NOT APPLE)
             target_link_options(${name} PRIVATE "LINKER:-z,relro,-z,now")
+            if(USE_GCOV)
+                if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+                    message(
+                        FATAL_ERROR "To use gcov, the build type must be Debug")
+                endif()
+                target_link_libraries(${name} PRIVATE --coverage)
```
Comment (lukaszstolarczuk) [inline cmake/helpers.cmake:135]: not `target_link_options`?

File: README.md [hunk 6]
```diff
@@ -123,6 +123,7 @@ List of options provided by CMake:
 | USE_TSAN | Enable ThreadSanitizer checks | ON/OFF | OFF |
 | USE_MSAN | Enable MemorySanitizer checks | ON/OFF | OFF |
 | USE_VALGRIND | Enable Valgrind instrumentation | ON/OFF | OFF |
+| USE_GCOV | Enable gcov support (Linux only) | ON/OFF | OFF |
```
Comment (lukaszstolarczuk) [inline README.md:126]: not sure how much we want to care at this point about other compilers, but will this work on gcc or other compilers as well? at least we can describe this somewhere - e.g. in your "example flow", you could add there extra flags `-DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++`

Comment (KFilipek) [reply README.md:126]: It works on gcc and clang

File: CONTRIBUTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (KFilipek) [reply CONTRIBUTING.md:206]: Done

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -202,3 +202,18 @@ New dependency: dependency_name
 license: SPDX license tag
 origin: https://dependency_origin.com
 ```
+## Code coverage
+
+After adding new functionality add tests and check coverage before and after change.
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply CONTRIBUTING.md:208]: Done

[the remark this answers is not in the document]
Comment (KFilipek) [reply CONTRIBUTING.md:208]: Done

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -202,3 +202,18 @@ New dependency: dependency_name
...
+
+```bash
+$ cmake -B build -DUSE_GCOV=1 -DCMAKE_BUILD_TYPE=Debug
+$ cmake --build build -j
+$ cd build
+$ ctest
+$ apt install lcov
+$ lcov --capture --directory . --output-file coverage.info
+$ genhtml -o html_report coverage.info
+```
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply CONTRIBUTING.md:221]: I will add a nightly with this change, then I will upgrade the readme.

File: cmake/helpers.cmake (hunk 5 above)
Comment (KFilipek) [reply cmake/helpers.cmake:135]: right

**Review by bratpiorka — APPROVED** [review]

File: cmake/helpers.cmake (hunk 4 above)
Comment (lukaszstolarczuk) [reply cmake/helpers.cmake:109]: Soo.. it might work in the build type "RelWithDebInfo"...?

**Comment by lukaszstolarczuk** [comment]
pls don't treat my last issue as a blocker 😉

**Pull request merged**